airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java

```diff
@@ -104,101 +118,249 @@ public AirbyteConnectionStatus check(JsonNode config) {
-              .collect(Collectors.toList()));
+          .withStreams(getTables(
+              database,
+              Optional.ofNullable(config.get("database")).map(JsonNode::asText),
+              Optional.ofNullable(config.get("schema")).map(JsonNode::asText))
+                  .stream()
+                  .map(t -> CatalogHelpers.createAirbyteStream(t.getName(), t.getFields())
+                      .withSupportedSyncModes(Lists.newArrayList(SyncMode.FULL_REFRESH, SyncMode.INCREMENTAL)))
+                  .collect(Collectors.toList()));
     }
   }
 
-  protected List<String> getExcludedInternalSchemas() {
-    return List.of();
-  }
+  @Override
+  public Stream<AirbyteMessage> read(JsonNode config, ConfiguredAirbyteCatalog catalog, JsonNode state) throws Exception {
+    final JdbcStateManager stateManager = new JdbcStateManager(state == null ? new JdbcState() : Jsons.object(state, JdbcState.class), catalog);
+    final Instant now = Instant.now();
 
-  protected List<TableInfo> getTables(final Database database) throws Exception {
-    return discoverInternal(database).stream()
-        .map(t -> {
-          final List<Field> fields = Arrays.stream(t.fields())
-              .map(f -> Field.of(f.getName(), jooqDataTypeToJsonSchemaType(f.getDataType())))
-              .collect(Collectors.toList());
+    final JdbcDatabase database = createDatabase(config);
+
+    final Map<String, TableInfoInternal> tableNameToTable = discoverInternal(
+        database,
+        Optional.ofNullable(config.get("database")).map(JsonNode::asText),
+        Optional.ofNullable(config.get("schema")).map(JsonNode::asText))
+            .stream()
+            .collect(Collectors.toMap(t -> String.format("%s.%s", t.getSchemaName(), t.getName()), Function.identity()));
+
+    return database.query(connection -> {
+      Stream<AirbyteMessage> resultStream = Stream.empty();
+
+      for (final ConfiguredAirbyteStream airbyteStream : catalog.getStreams()) {
+        final String streamName = airbyteStream.getStream().getName();
+        if (!tableNameToTable.containsKey(streamName)) {
+          LOGGER.info("Skipping stream {} because it is not in the source", streamName);
+          continue;
+        }
+
+        final TableInfoInternal table = tableNameToTable.get(streamName);
+        final Stream<AirbyteMessage> stream = getAirbyteStreamReadStream(
+            connection,
+            airbyteStream,
+            table,
+            stateManager,
+            now);
+        resultStream = Stream.concat(resultStream, stream);
+      }
+      return resultStream.onClose(() -> Exceptions.toRuntime(database::close));
+    });
+  }
 
-          return new TableInfo(String.format("%s.%s", t.getSchema().getName(), t.getName()), fields);
-        })
+  // get the read stream for an airbyte stream. the naming is accurate if unfortunate.
+  private Stream<AirbyteMessage> getAirbyteStreamReadStream(Connection connection,
```

Comment:
Why not call it: `createReadStream(ConfiguredAirbyteStream airbyteStream, Connection....)` that's a little less mouthful 😶

Reply:
hahaha. thanks. as you can tell the naming was a bit uninspired here.